File: SDK/dx_rt/test/release/validation_test/test_app/python/generator.py
"""
Generator class for creating test options with IE reuse optimization
Ported from C++ version
"""

import json
import os
import random
from pathlib import Path
from typing import List, Dict, Any, Tuple
from dataclasses import dataclass
import sys

from dx_engine import DeviceStatus
from dx_engine import InferenceOption

@dataclass
class IEOption:
    """Options for IE creation (model~ieOption from JSON)"""
    model_path: str
    dynamic_cpu_offloading: str  # "on" or "off"
    thread_type: str            # "single-ie" or "multi-ie"
    thread_count: int           # thread count
    ort: bool                   # true or false
    bound: str                  # "NPU_ALL", "NPU_0", "NPU_12"
    device: str                 # "all", "1", "0,1"

@dataclass
class ExecutionOption:
    """Options for execution (inferenceFunction, inoutOption from JSON)"""
    inference_function: str     # "sync", "async", "batch"
    input_style: str           # "single" or "multi"
    output_buffer: str         # "user" or "internal"
    async_method: str          # "callback" or "wait"
    callback_delay: int        # delay value
    loop: int                  # loop count
    time: int                  # time value
    bitmatch: bool             # bitmatch enabled

@dataclass
class TestCase:
    """Test case that combines one IE option with multiple execution options"""
    ie_option: IEOption
    exec_options: List[ExecutionOption]

def is_valid_device_option(device_option: str, num_devices: int) -> bool:
    """Helper function to validate device option"""
    if device_option == "all":
        return True
    
    # Parse comma-separated device IDs
    try:
        segments = device_option.split(',')
        for segment in segments:
            segment = segment.strip()
            if not segment.isdigit():
                return False
            device_id = int(segment)
            if device_id < 0 or device_id >= num_devices:
                return False
        return True
    except:
        return False

class Generator:
    """Generator class for creating test options with IE reuse optimization"""
    
    def __init__(self, base_path: str, json_path: str, random_mode: bool):
        self.base_path = base_path
        self.json_path = json_path
        self.random_mode = random_mode
        self.model_paths: List[str] = []
        self.test_cases: List[TestCase] = []
        self.json_document: Dict[str, Any] = {}

    def load_json(self) -> bool:
        """Load and parse JSON file"""
        try:
            with open(self.json_path, 'r') as f:
                self.json_document = json.load(f)
        except Exception as e:
            print(f"Error loading JSON file: {e}")
            return False

        # Extract model directory paths
        self.model_paths.clear()
        
        if "model" in self.json_document and isinstance(self.json_document["model"], list):
            for model_dir in self.json_document["model"]:
                if isinstance(model_dir, str):
                    full_path = os.path.join(self.base_path, model_dir)
                    dxnn_file = self._find_dxnn_file_in_directory(full_path)
                    if dxnn_file:
                        self.model_paths.append(dxnn_file)
                    else:
                        print(f"Warning: No .dxnn file found in {full_path}")
        else:
            print("Error: 'model' array not found in JSON")
            return False

        return True

    def _find_dxnn_file_in_directory(self, directory_path: str) -> str:
        """Find .dxnn file in the given directory"""
        try:
            for filename in os.listdir(directory_path):
                if filename.endswith('.dxnn'):
                    file_path = os.path.join(directory_path, filename)
                    if os.path.isfile(file_path):
                        return file_path
        except Exception as e:
            print(f"Error accessing directory {directory_path}: {e}")
        
        return ""

    def generate_test_cases(self):
        """Generate combined test cases (IE options + execution options)"""
        ie_options = self._generate_ie_options(self.json_document, self.model_paths)
        exec_options = self._generate_execution_options(self.json_document)

        self.test_cases = []
        for ie_option in ie_options:
            test_case = TestCase(ie_option=ie_option, exec_options=exec_options)
            self.test_cases.append(test_case)

        print(f"Generated {len(self.test_cases)} test cases")

    def _generate_ie_options(self, config: Dict[str, Any], model_paths: List[str]) -> List[IEOption]:
        """Generate IE options (model~ieOption combinations)"""
        ie_options = []
        
        # Get available device count for validation
        num_devices = 0
        try:
            num_devices = DeviceStatus.get_device_count()
        except Exception as e:
            print(f"Warning: Could not get device count: {e}")
            num_devices = -1 

        # Extract options using helper functions
        dyn_cpu_options = self._extract_string_array(config, ["configuration", "dynamic-cpu-offloading"])
        ort_options = self._extract_bool_array(config, ["ieOption", "ort"])
        bound_options = self._extract_string_array(config, ["ieOption", "bound"])
        device_options = self._extract_string_array(config, ["ieOption", "device"])

        # Extract thread style options
        thread_styles = []
        if "threadStyle" in config:
            for thread_style in config["threadStyle"]:
                if isinstance(thread_style, dict):
                    thread_type = thread_style.get("type", "single-ie")
                    thread_count = thread_style.get("count", 1)
                    thread_styles.append((thread_type, thread_count))
        else:
            # Default thread style if not specified
            thread_styles.append(("single-ie", 1))

        # Generate all combinations or random selection
        for model_path in model_paths:
            for dyn_cpu in dyn_cpu_options:
                for ort in ort_options:
                    for bound in bound_options:
                        for device in device_options:
                            # Validate device option
                            if not is_valid_device_option(device, num_devices):
                                print(f"Warning: Invalid device option '{device}' for {num_devices} devices")
                                continue

                            for thread_type, thread_count in thread_styles:
                                ie_option = IEOption(
                                    model_path=model_path,
                                    dynamic_cpu_offloading=dyn_cpu,
                                    thread_type=thread_type,
                                    thread_count=thread_count,
                                    ort=ort,
                                    bound=bound,
                                    device=device
                                )
                                ie_options.append(ie_option)

        # Randomize if requested
        if self.random_mode and ie_options:
            random.shuffle(ie_options)
            # Take a random subset (e.g., 50% of total)
            subset_size = max(1, len(ie_options) // 2)
            ie_options = ie_options[:subset_size]

        return ie_options

    def _generate_execution_options(self, config: Dict[str, Any]) -> List[ExecutionOption]:
        """Generate execution options from JSON"""
        exec_options = []
        
        # Extract execution options
        infer_funcs = self._extract_string_array(config, ["inferenceFunction"])
        input_styles = self._extract_string_array(config, ["inoutOption", "inputStyle"])
        output_buffers = self._extract_string_array(config, ["inoutOption", "outputBuffer"])
        async_methods = self._extract_string_array(config, ["inoutOption", "asyncMethod"])
        callback_delays = self._extract_int_array(config, ["inoutOption", "callbackDelay"])
        loops = self._extract_int_array(config, ["inoutOption", "loop"])
        times = self._extract_int_array(config, ["inoutOption", "time"])
        bitmatches = self._extract_bool_array(config, ["inoutOption", "bitmatch"])

        # Generate combinations
        for infer_func in infer_funcs:
            for input_style in input_styles:
                for output_buffer in output_buffers:
                    for async_method in async_methods:
                        for callback_delay in callback_delays:
                            for loop in loops:
                                for time in times:
                                    for bitmatch in bitmatches:
                                        exec_option = ExecutionOption(
                                            inference_function=infer_func,
                                            input_style=input_style,
                                            output_buffer=output_buffer,
                                            async_method=async_method,
                                            callback_delay=callback_delay,
                                            loop=loop,
                                            time=time,
                                            bitmatch=bitmatch
                                        )
                                        exec_options.append(exec_option)

        # Randomize if requested
        if self.random_mode and exec_options:
            random.shuffle(exec_options)
            # Take a random subset
            subset_size = max(1, len(exec_options) // 2)
            exec_options = exec_options[:subset_size]

        return exec_options

    def _extract_string_array(self, config: Dict[str, Any], keys: List[str]) -> List[str]:
        """Helper function to extract string array from nested JSON"""
        current = config
        try:
            for key in keys:
                current = current[key]
            if isinstance(current, list):
                return [str(item) for item in current if isinstance(item, (str, int, float))]
            else:
                return [str(current)]
        except (KeyError, TypeError):
            return []

    def _extract_bool_array(self, config: Dict[str, Any], keys: List[str]) -> List[bool]:
        """Helper function to extract bool array from nested JSON"""
        current = config
        try:
            for key in keys:
                current = current[key]
            if isinstance(current, list):
                return [bool(item) for item in current]
            else:
                return [bool(current)]
        except (KeyError, TypeError):
            return []

    def _extract_int_array(self, config: Dict[str, Any], keys: List[str]) -> List[int]:
        """Helper function to extract int array from nested JSON"""
        current = config
        try:
            for key in keys:
                current = current[key]
            if isinstance(current, list):
                return [int(item) for item in current if isinstance(item, (int, float))]
            else:
                return [int(current)]
        except (KeyError, TypeError):
            return []

    def get_test_cases(self) -> List[TestCase]:
        """Get generated test cases"""
        return self.test_cases

    def print_test_cases(self):
        """Print test cases for debugging"""
        if not self.test_cases:
            print("No test cases generated")
            return

        print("=== Generated Test Cases Summary ===")
        print(f"Total Test Cases: {len(self.test_cases)}")
        
        for i, test_case in enumerate(self.test_cases):
            print(f"\n--- Test Case {i+1} ---")
            ie_opt = test_case.ie_option
            print(f"Model: {ie_opt.model_path}")
            print(f"Dynamic CPU Offloading: {ie_opt.dynamic_cpu_offloading}")
            print(f"Thread Type: {ie_opt.thread_type}")
            print(f"Thread Count: {ie_opt.thread_count}")
            print(f"ORT: {ie_opt.ort}")
            print(f"Bound: {ie_opt.bound}")
            print(f"Device: {ie_opt.device}")
            print(f"Execution Options: {len(test_case.exec_options)}")

    def check_for_duplicates(self):
        """Check for duplicate test cases and execution options"""
        print("========== DUPLICATE CHECK ==========")
        
        # Check for duplicate IE options
        ie_option_groups = {}
        for i, test_case in enumerate(self.test_cases):
            ie_key = (
                test_case.ie_option.model_path,
                test_case.ie_option.dynamic_cpu_offloading,
                test_case.ie_option.thread_type,
                test_case.ie_option.thread_count,
                test_case.ie_option.ort,
                test_case.ie_option.bound,
                test_case.ie_option.device
            )
            
            if ie_key not in ie_option_groups:
                ie_option_groups[ie_key] = []
            ie_option_groups[ie_key].append(i)
        
        # Report duplicate IE options
        duplicate_ie_groups = 0
        for ie_key, indices in ie_option_groups.items():
            if len(indices) > 1:
                duplicate_ie_groups += 1
                print(f"[DUPLICATE IE] Group {duplicate_ie_groups}: TestCases {indices}")
                print(f"  Model: {ie_key[0]}")
                print(f"  Dynamic CPU Offloading: {ie_key[1]}")
                print(f"  Thread: {ie_key[2]} ({ie_key[3]})")
                print(f"  ORT: {ie_key[4]}")
                print(f"  Bound: {ie_key[5]}")
                print(f"  Device: {ie_key[6]}")
        
        # Check for duplicate execution options within each test case
        duplicate_exec_groups = 0
        for i, test_case in enumerate(self.test_cases):
            exec_option_counts = {}
            for j, exec_option in enumerate(test_case.exec_options):
                exec_key = (
                    exec_option.inference_function,
                    exec_option.input_style,
                    exec_option.output_buffer,
                    exec_option.async_method,
                    exec_option.callback_delay,
                    exec_option.loop,
                    exec_option.time,
                    exec_option.bitmatch
                )
                
                if exec_key not in exec_option_counts:
                    exec_option_counts[exec_key] = []
                exec_option_counts[exec_key].append(j)
            
            # Report duplicates in this test case
            for exec_key, indices in exec_option_counts.items():
                if len(indices) > 1:
                    duplicate_exec_groups += 1
                    print(f"[DUPLICATE EXEC] TestCase {i}, ExecOptions {indices}")
                    print(f"  Function: {exec_key[0]}")
                    print(f"  Input: {exec_key[1]}")
                    print(f"  Output: {exec_key[2]}")
                    print(f"  Async: {exec_key[3]}")
                    print(f"  Loop/Time: {exec_key[5]}/{exec_key[6]}")
                    print(f"  Bitmatch: {exec_key[7]}")
        
        # Summary
        print(f"========== DUPLICATE SUMMARY ==========")
        print(f"Duplicate IE Option Groups: {duplicate_ie_groups}")
        print(f"Duplicate Execution Option Groups: {duplicate_exec_groups}")
        print("=========================================")

def set_inference_configuration_from_ie_option(inference_option, ie_option: IEOption):
    """Set inference configuration from IE option
    This function needs to be implemented with actual DXRT Python API
    """

    # Set ORT option (now boolean)
    inference_option.useORT = ie_option.ort
    
    if ie_option.bound == "NPU_ALL":
        inference_option.boundOption = InferenceOption.BOUND_OPTION.NPU_ALL
    
    elif ie_option.bound == "NPU_0":
        inference_option.boundOption = InferenceOption.BOUND_OPTION.NPU_0
    
    elif ie_option.bound == "NPU_1":

        inference_option.boundOption = InferenceOption.BOUND_OPTION.NPU_1

    elif ie_option.bound == "NPU_2":
        inference_option.boundOption = InferenceOption.BOUND_OPTION.NPU_2

    elif ie_option.bound == "NPU_01" or ie_option.bound == "NPU_0/1":
        inference_option.boundOption = InferenceOption.BOUND_OPTION.NPU_01;  # NPU_0/1
    
    elif ie_option.bound == "NPU_12" or ie_option.bound == "NPU_1/2":
        inference_option.boundOption = InferenceOption.BOUND_OPTION.NPU_12;  # NPU_1/2

    elif ie_option.bound == "NPU_02" or ie_option.bound == "NPU_0/2":
        inference_option.boundOption = InferenceOption.BOUND_OPTION.NPU_02;  # NPU_0/2

    else:
        print(f"Error: Invalid bound option: {ie_option.bound}")
        sys.exit(-1)
    

    # Set device options
    inference_option.devices = [];  # Clear any existing devices
    
    if ie_option.device == "all":
        # Leave devices empty for "all" - engine will use all available devices
        pass
    
    else:
        device_options = ie_option.device.split(',')

        for device_str in device_options:
            device_str = device_str.strip()

            if device_str.isdigit():
                device_id = int(device_str)
                inference_option.devices.append(device_id)
                has_valid_device = True

            else:
                print(f"Error: Invalid device option: {ie_option.device} (contains non-numeric value: '{device_str}')")
                sys.exit(-1)
        
        # If no valid devices were parsed, it's an error
        if not has_valid_device:
            print(f"Error: Invalid device option: {ie_option.device} (no valid device IDs found)")
            sys.exit(-1)
        
    

File: SDK/dx_rt/test/release/validation_test/test_app/cpp/src/generator.cpp
#include "../include/generator.h"
#include "../include/utils.h"
#include "dxrt/dxrt_api.h"
#include <iostream>
#include <cstdlib>
#include <map>
#include <sstream>
#include <algorithm>
#include <fstream>
#include "dxrt/device_info_status.h"
#include "dxrt/extern/rapidjson/document.h"
#include "dxrt/extern/rapidjson/istreamwrapper.h"
#include "dxrt/extern/rapidjson/error/en.h"

using std::string;
using std::cout;
using std::endl;

// Helper function to validate device option
bool isValidDeviceOption(const string& deviceOption, int numDevices)
{
    if (deviceOption == "all")
    {
        return true; // "all" is always valid
    }
    
    // Parse comma-separated device IDs
    std::stringstream ss(deviceOption);
    string segment;
    
    while (std::getline(ss, segment, ','))
    {
        // Remove whitespace
        segment.erase(std::remove_if(segment.begin(), segment.end(), ::isspace), segment.end());
        if (segment.empty()) continue;
        
        try
        {
            int device_id = std::stoi(segment);
            if (device_id >= numDevices)
            {
                return false;
            }
        }
        catch (const std::exception& e)
        {
            cout << "Warning: Invalid device ID '" << segment << "' in device option: " << deviceOption << endl;
            return false;
        }
    }
    
    return true;
}

Generator::Generator(const std::string& basePath, const std::string& jsonPath, bool random)
    : _basePath(basePath), _jsonPath(jsonPath), _random(random)
{

}

Generator::~Generator()
{

}

bool Generator::LoadJson()
{
    // Read and parse JSON file
    std::ifstream json_stream(_jsonPath);
    if (!json_stream.is_open()) 
    {
        cout << "Error: Unable to open json file: " << _jsonPath << endl;
        return false;
    }

    rapidjson::Document jsonConfig;
    rapidjson::IStreamWrapper isw(json_stream);
    jsonConfig.ParseStream(isw);
    json_stream.close();
    
    if (jsonConfig.HasParseError())
    {
        cout << "Error: Failed to parse json file - " << rapidjson::GetParseError_En(jsonConfig.GetParseError()) << endl;
        return false;
    }

    // Extract model directory paths (will find .dxnn files in generator)
    _modelPaths.clear();
    
    if (jsonConfig.HasMember("model") && jsonConfig["model"].IsArray())
    {
        const rapidjson::Value& models = jsonConfig["model"];
        for (rapidjson::SizeType i = 0; i < models.Size(); i++)
        {
            if (models[i].IsString())
            {
                string directory_path = _basePath;
                if (!directory_path.empty() && directory_path.back() != '/')
                {
                    directory_path += "/";
                }
                directory_path += models[i].GetString();
                _modelPaths.push_back(directory_path);
            }
        }
    }
    else
    {
        cout << "Error: 'model' field not found or not an array in json file" << endl;
        return false;
    }

    // Store the parsed JSON document for later use in GenerateTestCases
    _jsonDocument.CopyFrom(jsonConfig, _jsonDocument.GetAllocator());
    
    return true;
}

void Generator::PrintTestCases()
{
    // Check if test cases are initialized
    if (_testCases.empty())
    {
        cout << "Test cases not initialized. Please call LoadJson() and GenerateTestCases() first." << endl;
        return;
    }

    cout << "=== Generated Test Cases Summary ===" << endl;
    cout << "Total Test Cases: " << _testCases.size() << endl << endl;

    for (size_t i = 0; i < _testCases.size(); i++)
    {
        const TestCase& testCase = _testCases[i];
        
        cout << "=== Test Case " << (i + 1) << " ===" << endl;
        
        // Print IE Options
        cout << "IE Options:" << endl;
        cout << "  Model: " << testCase.ieOption.model_path << endl;
        cout << "  Dynamic CPU Offloading: " << testCase.ieOption.dynamicCpuOffloading << endl;
        cout << "  Thread Type: " << testCase.ieOption.threadType << endl;
        cout << "  Thread Count: " << testCase.ieOption.threadCount << endl;
        cout << "  ORT: " << (testCase.ieOption.ort ? "true" : "false") << endl;
        cout << "  Bound: " << testCase.ieOption.bound << endl;
        cout << "  Device: " << testCase.ieOption.device << endl;
        
        // Print Execution Options
        cout << endl << "Execution Options (" << testCase.execOptions.size() << " options):" << endl;
        for (size_t j = 0; j < testCase.execOptions.size(); j++)
        {
            const ExecutionOption& execOpt = testCase.execOptions[j];
            cout << "  " << (j + 1) << ". Function: " << execOpt.inferenceFunction
                 << ", Input: " << execOpt.inputStyle
                 << ", Output: " << execOpt.outputBuffer
                 << ", Async: " << execOpt.asyncMethod
                 << ", Loop: " << execOpt.loop
                 << ", Time: " << execOpt.time
                 << ", Bitmatch: " << (execOpt.bitmatch ? "true" : "false") << endl;
        }
        
        cout << "------------------------" << endl << endl;
    }
}

void Generator::CheckForDuplicates() const
{
    if (_testCases.empty())
    {
        cout << "No test cases to check for duplicates." << endl;
        return;
    }
    
    cout << "=== Checking for Duplicate Test Cases ===" << endl;
    
    int duplicateCount = 0;
    std::vector<bool> isDuplicate(_testCases.size(), false);
    
    // Check for duplicate IE options
    std::map<IEOption, std::vector<size_t>> ieOptionMap;
    for (size_t i = 0; i < _testCases.size(); i++)
    {
        ieOptionMap[_testCases[i].ieOption].push_back(i);
    }
    
    // Report IE option duplicates
    for (const auto& pair : ieOptionMap)
    {
        if (pair.second.size() > 1)
        {
            cout << "Duplicate IE Option found in test cases: ";
            for (size_t idx : pair.second)
            {
                cout << (idx + 1) << " ";
                isDuplicate[idx] = true;
            }
            cout << endl;
            cout << "  Model: " << pair.first.model_path << endl;
            cout << "  Dynamic CPU: " << pair.first.dynamicCpuOffloading << endl;
            cout << "  Thread Type: " << pair.first.threadType << endl;
            cout << "  Thread Count: " << pair.first.threadCount << endl;
            cout << "  ORT: " << (pair.first.ort ? "true" : "false") << endl;
            cout << "  Bound: " << pair.first.bound << endl;
            cout << "  Device: " << pair.first.device << endl;
            cout << endl;
            duplicateCount++;
        }
    }
    
    // Check for duplicate execution options within each test case
    for (size_t i = 0; i < _testCases.size(); i++)
    {
        const auto& execOptions = _testCases[i].execOptions;
        std::map<ExecutionOption, std::vector<size_t>> execOptionMap;
        
        for (size_t j = 0; j < execOptions.size(); j++)
        {
            execOptionMap[execOptions[j]].push_back(j);
        }
        
        for (const auto& pair : execOptionMap)
        {
            if (pair.second.size() > 1)
            {
                cout << "Duplicate Execution Options found in test case " << (i + 1) << " at positions: ";
                for (size_t idx : pair.second)
                {
                    cout << (idx + 1) << " ";
                }
                cout << endl;
                cout << "  Function: " << pair.first.inferenceFunction << endl;
                cout << "  Input Style: " << pair.first.inputStyle << endl;
                cout << "  Output Buffer: " << pair.first.outputBuffer << endl;
                cout << "  Async Method: " << pair.first.asyncMethod << endl;
                cout << "  Loop: " << pair.first.loop << endl;
                cout << "  Time: " << pair.first.time << endl;
                cout << "  Bitmatch: " << (pair.first.bitmatch ? "true" : "false") << endl;
                cout << endl;
                duplicateCount++;
            }
        }
    }
    
    if (duplicateCount == 0)
    {
        cout << "No duplicates found in test cases." << endl;
    }
    else
    {
        cout << "Total duplicate groups found: " << duplicateCount << endl;
    }
    
    cout << "=== Duplicate Check Complete ===" << endl;
}

// Generate IE options (model~ieOption combinations)
std::vector<IEOption> Generator::generateIEOptions(const rapidjson::Value& config, const std::vector<string>& modelPaths)
{
    std::vector<IEOption> ie_options;
    
    // Get available device count for validation
    int num_devices = 0;

    try 
    {
        num_devices = dxrt::DeviceStatus::GetDeviceCount();
    }
    catch (const std::exception& e)
    {
        cout << "Warning: Failed to get device count, proceeding without device validation: " << e.what() << endl;
        num_devices = -1; // Disable validation
    }
    
    // Extract options using helper functions
    auto dyn_cpu_options = extractStringArray(config, {"configuration", "dynamic-cpu-offloading"});
    auto ort_options = extractBoolArray(config, {"ieOption", "ort"});
    auto bound_options = extractStringArray(config, {"ieOption", "bound"});
    auto device_options = extractStringArray(config, {"ieOption", "device"});

    // Extract thread style options (array of objects)
    struct ThreadStyle { string type; int count; };
    std::vector<ThreadStyle> thread_styles;
    
    if (config.HasMember("threadStyle") && config["threadStyle"].IsArray())
    {
        const rapidjson::Value& threadStyleArray = config["threadStyle"];
        thread_styles.reserve(threadStyleArray.Size());
        for (rapidjson::SizeType i = 0; i < threadStyleArray.Size(); i++)
        {
            const rapidjson::Value& style = threadStyleArray[i];
            if (style.HasMember("type") && style.HasMember("count") && 
                style["type"].IsString() && style["count"].IsInt())
            {
                thread_styles.emplace_back(ThreadStyle{
                    style["type"].GetString(), 
                    style["count"].GetInt()
                });
            }
        }
    }
    
    // Calculate total combinations and reserve space for better performance
    size_t totalCombinations = modelPaths.size() * dyn_cpu_options.size() * thread_styles.size() * 
                              ort_options.size() * bound_options.size() * device_options.size();
    ie_options.reserve(totalCombinations);
    
    // Generate all combinations for IE options
    for (const auto& model_directory : modelPaths)
    {
        // Find .dxnn file in the directory
        string actual_model_path = findDxnnFileInDirectory(model_directory);
        if (actual_model_path.empty())
        {
            cout << "Error: No .dxnn file found in directory: " << model_directory << endl;
            continue; // Skip this model directory
        }

        if (_random)
        {
            // Run flow will slightly adjusted to maximiaze random option coverage
            for (const auto& dyn_cpu : dyn_cpu_options)
            {
                for (const auto& ort : ort_options)
                {
                    for (const auto& thread_style : thread_styles)
                    {
                        auto bound = GetRandomElement(bound_options);
                        auto device = GetRandomElement(device_options);

                        // Validate device option against available devices
                        if (num_devices > 0 && !isValidDeviceOption(device, num_devices))
                        {
                            continue; // Skip invalid device option
                        }

                        // Pick Random options
                        auto tmp_count = GetRandomInt(thread_style.count, 1);
                        
                        ie_options.emplace_back(IEOption{
                            actual_model_path, dyn_cpu, thread_style.type, tmp_count,
                            ort, bound, device
                        });
                    }
                }
            }
        }
        else
        {
            for (const auto& dyn_cpu : dyn_cpu_options)
            {
                for (const auto& thread_style : thread_styles)
                {
                    for (const auto& ort : ort_options)
                    {
                        for (const auto& bound : bound_options)
                        {
                            for (const auto& device : device_options)
                            {
                                // Validate device option against available devices
                                if (num_devices > 0 && !isValidDeviceOption(device, num_devices))
                                {
                                    continue; // Skip invalid device option
                                }
                                
                                ie_options.emplace_back(IEOption{
                                    actual_model_path, dyn_cpu, thread_style.type, thread_style.count,
                                    ort, bound, device
                                });
                            }
                        }
                    }
                }
            }
        }
    }
    
    return ie_options;
}

// Generate execution options (inferenceFunction, inoutOption combinations from JSON)
std::vector<ExecutionOption> Generator::generateExecutionOptions(const rapidjson::Value& config)
{
    std::vector<ExecutionOption> exec_options;
    
    // Extract options using helper functions
    auto inference_function_options = extractStringArray(config, {"inferenceFunction"});
    auto input_style_options = extractStringArray(config, {"inoutOption", "inputStyle"});
    auto output_buffer_options = extractStringArray(config, {"inoutOption", "outputBuffer"});
    auto async_method_options = extractStringArray(config, {"inoutOption", "asyncMethod"});
    
    // Extract fixed values from inoutOption
    int callback_delay = config["inoutOption"]["callbackDelay"].GetInt();
    int loop = config["inoutOption"]["loop"].GetInt();
    int time = config["inoutOption"]["time"].GetInt();
    bool bitmatch = config["inoutOption"]["bitmatch"].GetBool();
    
    // Reserve space for better performance
    size_t total_combinations = inference_function_options.size() * input_style_options.size() * 
                              output_buffer_options.size() * async_method_options.size();
    exec_options.reserve(total_combinations);
    
    // Generate all combinations for execution options
    for (const auto& inference_func : inference_function_options)
    {
        for (const auto& input_style : input_style_options)
        {
            for (const auto& output_buffer : output_buffer_options)
            {
                for (const auto& async_method : async_method_options)
                {
                    exec_options.emplace_back(ExecutionOption{
                        inference_func, input_style, output_buffer, async_method,
                        callback_delay, loop, time, bitmatch
                    });
                }
            }
        }
    }
    
    return exec_options;
}

// Generate combined test cases (IE options + execution options)
void Generator::GenerateTestCases()
{
    _testCases.clear();
    
    auto ie_options = generateIEOptions(_jsonDocument, _modelPaths);
    auto exec_options = generateExecutionOptions(_jsonDocument);
    
    // Create one test case per unique IE option, with all execution options
    _testCases.reserve(ie_options.size());
    
    for (const auto& ie_option : ie_options)
    {
        TestCase test_case;
        test_case.ieOption = ie_option;
        test_case.execOptions = exec_options; // Store all execution options for this IE
        _testCases.push_back(test_case);
    }
}

// Helper function implementations
std::vector<string> Generator::extractStringArray(const rapidjson::Value& config, const std::vector<string>& keys)
{
    std::vector<string> result;
    const rapidjson::Value* current = &config;
    
    // Navigate through nested keys
    for (const auto& key : keys)
    {
        if (current->HasMember(key.c_str()))
        {
            current = &(*current)[key.c_str()];
        }
        else
        {
            return result; // Return empty if key not found
        }
    }
    
    // Extract array values
    if (current->IsArray())
    {
        for (rapidjson::SizeType i = 0; i < current->Size(); i++)
        {
            const rapidjson::Value& opt = (*current)[i];
            if (opt.IsString())
            {
                result.push_back(opt.GetString());
            }
        }
    }
    
    return result;
}

std::vector<bool> Generator::extractBoolArray(const rapidjson::Value& config, const std::vector<string>& keys)
{
    std::vector<bool> result;
    const rapidjson::Value* current = &config;
    
    // Navigate through nested keys
    for (const auto& key : keys)
    {
        if (current->HasMember(key.c_str()))
        {
            current = &(*current)[key.c_str()];
        }
        else
        {
            return result; // Return empty if key not found
        }
    }
    
    // Extract array values
    if (current->IsArray())
    {
        for (rapidjson::SizeType i = 0; i < current->Size(); i++)
        {
            const rapidjson::Value& opt = (*current)[i];
            if (opt.IsBool())
            {
                result.push_back(opt.GetBool());
            }
        }
    }
    
    return result;
}

std::vector<int> Generator::extractIntArray(const rapidjson::Value& config, const std::vector<string>& keys)
{
    std::vector<int> result;
    const rapidjson::Value* current = &config;
    
    // Navigate through nested keys
    for (const auto& key : keys)
    {
        if (current->HasMember(key.c_str()))
        {
            current = &(*current)[key.c_str()];
        }
        else
        {
            return result; // Return empty if key not found
        }
    }
    
    // Extract array values
    if (current->IsArray())
    {
        for (rapidjson::SizeType i = 0; i < current->Size(); i++)
        {
            const rapidjson::Value& opt = (*current)[i];
            if (opt.IsInt())
            {
                result.push_back(opt.GetInt());
            }
        }
    }
    
    return result;
}

void SetInferenceConfigurationFromIEOption(dxrt::InferenceOption& op, const IEOption& ieOption)
{
    // Set ORT option (now boolean)
    op.useORT = ieOption.ort;

    // Based on run_model.cpp, the bound option mapping:
    // 0: NPU_ALL, 1: NPU_0, 2: NPU_1, 3: NPU_2, 4: NPU_0/1, 5: NPU_1/2, 6: NPU_0/2
    
    if (ieOption.bound == "NPU_ALL")
    {
        op.boundOption = 0;
    }
    else if (ieOption.bound == "NPU_0")
    {
        op.boundOption = 1;
    }
    else if (ieOption.bound == "NPU_1")
    {
        op.boundOption = 2;
    }
    else if (ieOption.bound == "NPU_2")
    {
        op.boundOption = 3;
    }
    else if (ieOption.bound == "NPU_01" || ieOption.bound == "NPU_0/1")
    {
        op.boundOption = 4;  // NPU_0/1
    }
    else if (ieOption.bound == "NPU_12" || ieOption.bound == "NPU_1/2")
    {
        op.boundOption = 5;  // NPU_1/2
    }
    else if (ieOption.bound == "NPU_02" || ieOption.bound == "NPU_0/2")
    {
        op.boundOption = 6;  // NPU_0/2
    }
    else
    {
        cout << "Error: Invalid bound option: " << ieOption.bound << endl;
        std::exit(-1);
    }

    // Set device options
    op.devices.clear();  // Clear any existing devices
    
    if (ieOption.device == "all")
    {
        // Leave devices empty for "all" - engine will use all available devices
    }
    else
    {
        // Parse comma-separated device IDs
        std::stringstream ss(ieOption.device);
        string segment;
        bool has_valid_device = false;
        
        while (std::getline(ss, segment, ','))
        {
            // Remove whitespace
            segment.erase(std::remove_if(segment.begin(), segment.end(), ::isspace), segment.end());
            if (segment.empty()) continue;
            
            // Check if segment contains only digits
            if (segment.find_first_not_of("0123456789") != std::string::npos)
            {
                cout << "Error: Invalid device option: " << ieOption.device << " (contains non-numeric value: '" << segment << "')" << endl;
                std::exit(-1);
            }
            
            try
            {
                int device_id = std::stoi(segment);
                op.devices.push_back(device_id);
                has_valid_device = true;
            }
            catch (const std::exception& e)
            {
                cout << "Error: Invalid device option: " << ieOption.device << " (failed to parse: '" << segment << "')" << endl;
                std::exit(-1);
            }
        }
        
        // If no valid devices were parsed, it's an error
        if (!has_valid_device)
        {
            cout << "Error: Invalid device option: " << ieOption.device << " (no valid device IDs found)" << endl;
            std::exit(-1);
        }
    }
}
